### DISCUSSION TITLE
Wishlist - Import (selected) COM/ActiveX Library Constants/Enums/etc into project code

### DISCUSSION THREAD

**Discussion by mburns08109** [twinbasic/twinbasic#538] [Ideas]
This is more likely a request for the IDE than for the tB compiler itself, but in a NoLongerSet.com topic some comments were made about issues surrounding Early-vs-late binding and one solution to help was to reproduce the early-bound library's constants in our VBx code.
This led to my ruminating that it would be neat to never need to do that sort of thing again. After all, if the typlib is available for development, why can't the IDE (if not the compiler) do that FOR us? When it sees a reference to a COM/Ax library-defined constant or enum value in our code - just automatically IMPORT the relevant entries directly into our project - so that a later desire to switch to late-bound from early-bound dev modes of working can be more easily facilitated?
(Maybe you very bright dev guys have this in your queue already?)

**Comment by WaynePhillipsEA** [comment]
The plan on this is to have a build configuration option that allows you to generate late-bound compilations without having to change your  code.   So for example, if you write `New Access.Application` the compiler might generate `CreateObject("Access.Application")` for you, and `Dim myAccess As Access.Application` would be treated as `Dim myAccess As Object` etc.

By doing it the way described above, we can have all the benefits of early-binding at the IDE level (i.e. intellisense/signature-help), and all the benefits of late-binding at the compilation level... without having to change a single line of code.

**Reply by wqweto** [reply]
No need to change New Access.Application to become late-bound. Just using As Object for the interfaces is enough.

**Reply by WaynePhillipsEA** [reply]
CLSIDs/IIDs can be version dependent, so it is best to change to CreateObject to avoid CLSIDs/IIDs being relied upon.

**Reply by wqweto** [reply]
Is there any IID involved in New Access.Application other than IID_IUnknown?

**Reply by WaynePhillipsEA** [reply]
Yes, the correct IID (IID__Application) is passed to CoCreateInstance

**Reply by wqweto** [reply]
In this regard it seems Version independent <> Late-bound.

I just peeked at VB6 codegen for
```
    Dim cn1 As Object
    Dim cn2 As ADODB.Connection
        
    Set cn1 = New ADODB.Connection
    Set cn2 = New ADODB.Connection
```
. . . and it's not using `CoCreaterInstance[Ex]` at all but delegates to `__vbaNew` from runtime which directly uses `CoGetClassObject` to get `IClassFactory` (or `IClassFactory2` for licenced instances) and then calls `CreateInstance` passing `IID__Connection` for `riid` in both cases.

This prevents first instantiation above from being Version independent indeed which is unfortunate as the variable is dimensioned `As Object` and `IID_IDispatch` would have been enough to future-proof the instantiation.

**Reply by WaynePhillipsEA** [reply]
As an interesting aside; VBA (7 at least) uses CoCreateInstanceEx.

**Reply by mburns08109** [reply]
> The plan on this is to have a build configuration option that allows you to generate late-bound compilations without having to change your code. So for example, if you write `New Access.Application` the compiler might generate `CreateObject("Access.Application")` for you, and `Dim myAccess As Access.Application` would be treated as `Dim myAccess As Object` etc.
> 
> By doing it the way described above, we can have all the benefits of early-binding at the IDE level (i.e. intellisense/signature-help), and all the benefits of late-binding at the compilation level... without having to change a single line of code.

Sorry to be pedantic here, Wayne, but does that mean that you already planned on automatically bringing the Excel.xlVeryHidden definition into code somewhere as something akin to "Public Const xlHidden = 2 'from COM library entry Excel.xlVeryHidden" so that code-based references to xlVeryHidden work regardless of the early/late binding style?  ...and if so, how do you plan to auto-document the source for the reference (for versioning issues)?

**Reply by WaynePhillipsEA** [reply]
No, it's much simpler than that.  At the IDE level, the reference to Excel will still exist, so we don't have to import anything at all... and the constants from the Excel type library are all still available to use...

The build configuration switch will simply tell the compiler NOT to early-bind any objects during codegen, issuing late-bound calls at all appropriate points.

**Reply by mburns08109** [reply]
OK, I did not understand that.

What you're saying is that if, in my code, I have a line like:
oWB.Sheets("MyHiddenSheet").visible = Excel.xlVeryHidden  ' = 2
'oWB being an Excel workbook reference, of course

...that those constant or enum references sprinkled throughout the codebase will continue to function properly, when the binding changes from early to late - later in the build stage?!  How? Will the compiler change the code from a symbolic lookup/pointer to a literal constant during compilation (but without changing the source code)?

I'd like to see that, but color me skeptical...and doesn't that pose a debugging problem should later COM library versions change the underlying value for the constants/enums? (I'm not sure if those would be considered "contract-breaking interface changes" or not.)

**Reply by WaynePhillipsEA** [reply]
> ...that those constant or enum references sprinkled throughout the codebase will continue to function properly, when the binding changes from early to late - later in the build stage?! How? Will the compiler change the code from a symbolic lookup/pointer to a literal constant during compilation (but without changing the source code)?

There will be no change needed in the compiler for constants/enums... it still has the Excel reference, and so it can still see those symbols and bind to their literal values as it ordinarily does.

>I'd like to see that, but color me skeptical...and doesn't that pose a debugging problem should later COM library versions change the underlying value for the constants/enums? (I'm not sure if those would be considered "contract-breaking interface changes" or not.)

Only in the same way that importing the constant/enum values into your project would do the same.   That's just a limitation of using late-bound code in general.  In practice, it is rarely a problem, and so not a consideration for this feature.  

The only changes the proposed switch will make are:
For `New <class>` the compiler will generate `CreateObject("<class>")` for you, and `Dim myVar As <class/interface>` would be treated as `Dim myAccess As Object` where appropriate.    That's pretty much all that has to happen to make the feature work.

**Reply by mwolfe02** [reply]
Just so I make sure I understand, the reason this approach would work in twinBASIC but wouldn't work in VBA is that tB is *fully compiled* whereas VBA only "compiles" to P-Code.  In other words, the final "compilation" of VBA into machine instructions happens *on the end user's computer*, and so to have access to third-party constants (such as `Excel.xlVeryHidden`) at runtime in VBA you would be relying on that external reference being available on the end user's computer.  Since tB is fully compiled, the conversion of the code to machine instructions happens *on the developer's computer*.  Hence, the third-party constants can be embedded into the machine instructions at compile time on the developer's machine. 

Am I understanding correctly?

**Reply by WaynePhillipsEA** [reply]
Compiled P-code (or ExCode) also contains the resolved literal values.  But VBA often discards the compiled code persisted in a project and goes back to the source code for recompilation (assuming not an ACCDE/MDE).  In a compiled state, the constant values are reduced to their literal values, but VBA simply doesn't guarantee that compiled state, and so there are runtime dependencies on type libraries.

**Reply by mwolfe02** [reply]
Got it.  Thanks for clarifying when the constants get resolved in P-Code/ExCode.

**Reply by mburns08109** [reply]
> > ...that those constant or enum references sprinkled throughout the codebase will continue to function properly, when the binding changes from early to late - later in the build stage?! How? Will the compiler change the code from a symbolic lookup/pointer to a literal constant during compilation (but without changing the source code)?
> 
> There will be no change needed in the compiler for constants/enums... it still has the Excel reference, and so it can still see those symbols and bind to their literal values as it ordinarily does.
> 
> > I'd like to see that, but color me skeptical...and doesn't that pose a debugging problem should later COM library versions change the underlying value for the constants/enums? (I'm not sure if those would be considered "contract-breaking interface changes" or not.)
> 
> Only in the same way that importing the constant/enum values into your project would do the same. That's just a limitation of using late-bound code in general. In practice, it is rarely a problem, and so not a consideration for this feature.
> 
> The only changes the proposed switch will make are: For `New <class>` the compiler will generate `CreateObject("<class>")` for you, and `Dim myVar As <class/interface>` would be treated as `Dim myAccess As Object` where appropriate. That's pretty much all that has to happen to make the feature work.

@WaynePhillipsEA 
SO, I'm going to push for this request a bit more and say...See my back and forth with Eduardo. 

My simple reply to you, in the light of that discussion is: "this request is much less about how the compiler works, and much more about how we humans work (especially when given the passage of time)."

**Comment by EduardoVB** [comment]
> When it sees a reference to a COM/Ax library-defined constant or enum value in our code - just automatically IMPORT the relevant entries directly into our project

Suppose that you write:

`Dim MyRec As RecordSet`

What DLL should it reference, ADO or DAO?

**Reply by mburns08109** [reply]
Hmmm... varies with the project file's references I would think - and following on that - the sequencing of the references would have to govern. i.e. if the DAO library reference preceeds the ADO library reference in the list, then your unspecified recordset would be a DAO recordset by default. (OTOH, of the ADO library preceeds the DAO library in the list sequence, then it would be an ADO recordset.)

That would be consistent with VBA behavior, at least.

If you're one of THOSE programmers who doesn't specify precisely what they mean in order to avoid such confusion, well then ya gets what ya deserves if someone screws up the library references list for ya then.  :-)

**Reply by EduardoVB** [reply]
You proposed to automatically add the reference to the reference list (when they are still not referenced) based in some object type that you typed in code (if I understood you correctly).

Then, there is no list to look for. The "list" is the whole list of registered type libraries in the registry.

Then again, how would you know whether it refers to DAO or ADO (or something else) when you read "As Database"?.

**Reply by mburns08109** [reply]
Eduardo,

In VBx there is either a references list attached to the project, or in the .VBP file there is a references list.
It is these that I am referencing (as opposed to any general list in the GAC/registry) - and in both of those cases, the sequence of the listed references govern the resolution of unspecified object types. Hence whether an unspecified recordset object declaration/reference is interpreted at compile-time (or run-time) as either an ADO recordset or DAO recordset is governed by the sequence of the references list (presuming both are in the references list, of course).

As for your "Database" question, well the ADO doesn't have a "Database" object, so that's a DAO object reference since no other object library I'm thinking of has one of those, IIRC.

**Reply by EduardoVB** [reply]
I didn't remember that ADO doesn't have Database object because I use ADO very little.
But that wasn't the point, make it Recordset then.

You said in the OP:

> about issues surrounding Early-vs-late binding and one solution to help was to reproduce the early-bound library's constants in our VBx code.
> This led to my ruminating that it would be neat to never need to do that sort of thing again. After all, if the typlib is available for development, why can't the IDE (if not the compiler) do that FOR us? When it sees a reference to a COM/Ax library-defined constant or enum value in our code - just automatically IMPORT the relevant entries directly into our project

OK, you were talking about constants and enums, not objects. But the issue is the same.

Suppose that in your late bound code you want to write a constant, like `flexAlignCenterTop`
You don't have the MSFlexGrid, nor the MS**H**Flexgrid in your reference list, because you said you were working late bound. 
You want the compiler to automatically figure the value of `flexAlignCenterTop` by looking in some reference.
I repeat: you are late bound, you don't have any of these libraries referenced. If you had any of them referenced, then you already would have the constant declared, so you wouldn't need the compiler to figure anything.

For finding the value of that constant, the IDE (or compiler, whatever) would have to look in every available typelib registered in the registry.
In that particular case the value for that constant will be the same for MSFlexGrid and MSHFlexgrid, but they could differ in other cases (same names constant with different values from different libraries).

How would the compiler know from what library to take the constant value from?
(Please stop repeating "from the reference list" because when you are late bound the libraries are not in the reference list of the project, and if you had the library referenced in the project, then the constant would be already declared from the library and there is no point in what you are asking).

**Reply by mburns08109** [reply]
Eduardo,

The thesis underlying all this is "Develop while Early Bound, and Deploy as late bound"
Therefore, the early bound libraries (and references list) IS available - at some point in the Dev/Test/Deploy cycle, the developer throws the "Switch project to Late Bound mode" switch (functionally, at least).
This was the preface driving the request.

Our thinking at the time was: "Oh, well it'd probably be best to import all the used enums and constants defined in the referenced COM/ActiveX library directly into the source code *at that point* so that any future developers have the (now-static) enums and constants directly available in the source code" - for the sake of future changes/maintenance developers to see what was what when the original development/deployment was being completed.

If we failed to phrase it correctly so that intent was clear, I plead guilty to possessing inadequate communications skills.

Does this original request make more sense now?

Now, Wayne has explained how this is not necessary because of how tB works.
I (think I) understand his point, but I also think that "for future SDLC purposes", the intent of the request remains valid - due to the potential for COM/ActiveX library version changes over time to alter or eliminate certain details over time, making future maintenance more of an issue. For example, suppose a future version of the Excel ActiveX library redefines the xlVeryHidden constant from its present value to some new value.

If, at the time of project early-/late-bound conversion/compilation the then-current values were imported into the code, then Mr./Ms. FutureDeveloper would be able to see what was originally used and compare it to the current library value and react accordingly to make maintenance fixes.

Here's the scenario in the back of my mind on this: At some point, I KNOW that MS did something like this - An small set of Excel constants were altered/replaced with a set of bit-mapped constants (with new ones added), and in doing so, some pre-existing constant values were changed. That is the sort of "breaking change" I have in mind in making this request. This sort of "implementation detail" change is exactly the sort of information that tended NOT to be well communicated when Microsoft rolled out its "next great version of Office" in the past. This request is intended as a defense for THAT sort of issue.

**Reply by EduardoVB** [reply]
> Does this original request make more sense now?

Of course, and I think that it is a good idea.
Sorry that I misunderstood.

Reading again, and I think the Wayne in the[ first answer](https://github.com/WaynePhillipsEA/twinbasic/discussions/538#discussioncomment-1770414) suggested the best way to handle that: automatic conversion from early binding to late binding.

Suppose you are working with Word automation.
Now in VB6, if you work with early binding you have the advantage of the intellisense, but have the disadvantage that if the end user have installed other Office version, it won't work.
So the only "right" thing to do is to work late bound. And get used to work without intellisense.

But if we could get both advantages (intellisense and late bound), it would be great.
